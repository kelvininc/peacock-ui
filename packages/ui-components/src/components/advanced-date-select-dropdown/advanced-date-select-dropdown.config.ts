import { ComputePositionConfig, offset, size, MiddlewareState } from '@floating-ui/dom';
import { EIconName } from '../icon/icon.types';
import { EInputFieldType, EValidationState, ITextField } from '../text-field/text-field.types';
import { DEFAULT_DROPDOWN_Z_INDEX } from '../../globals/config';

const DEFAULT_DROPDOWN_OFFSET = 8;

// It needs to be lower because there are other dropdowns with a portal inside
export const ADVANCE_SELECT_PORTAL_Z_INDEX = DEFAULT_DROPDOWN_Z_INDEX - 1;

export const DEFAULT_DATE_INPUT_CONFIG: Partial<ITextField> = {
	placeholder: 'Select a date',
	type: EInputFieldType.Text,
	icon: EIconName.Calendar,
	state: EValidationState.None
};

export const DEFAULT_DROPDOWN_POSITION_OPTIONS: Partial<ComputePositionConfig> = {
	placement: 'bottom-end',
	middleware: [
		offset(DEFAULT_DROPDOWN_OFFSET),
		size({
			apply({ elements }: MiddlewareState) {
				Object.assign(elements.floating.style, {
					width: '544px'
				});
			}
		})
	]
};
