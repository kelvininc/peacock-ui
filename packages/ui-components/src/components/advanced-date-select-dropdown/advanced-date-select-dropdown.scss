@import '../../assets/styles/globals';

.advanced-date-select-dropdown {
	--dropdown-absolute-time-selected-input-width: #{365px};

	user-select: none;

	&--selected-absolute {
		kv-dropdown kv-text-field {
			--input-width: var(--dropdown-absolute-time-selected-input-width);
		}
	}
}

.advanced-date-select-calendar-container {
	display: inline-block;

	kv-calendar-advanced-date-selector::part(calendar) {
		border-bottom-left-radius: 0;
		border-bottom-right-radius: 0;
	}

	.footer {
		background-color: kv-color('neutral-7');
		border-color: kv-color('neutral-6');
		border-style: solid;
		border-width: 0 1px 1px;

		display: flex;
		justify-content: space-between;
		align-items: center;

		padding: $spacing-4x;

		border-bottom-left-radius: 4px;
		border-bottom-right-radius: 4px;

		.selected-time {
			color: kv-color('neutral-2');

			.label {
				@include kv-font-span-semibold;
				margin-bottom: $spacing;
			}

			.value {
				@include kv-font-label-small-regular;
			}
		}

		.actions {
			display: flex;
			gap: $spacing-3x;
		}
	}
}
