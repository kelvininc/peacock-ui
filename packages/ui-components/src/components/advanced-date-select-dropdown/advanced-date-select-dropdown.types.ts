import { ComputePositionConfig } from '@floating-ui/dom';
import { EventEmitter } from '@stencil/core';
import { ITextField, ICalendarAdvanceTimeChange, ICalendarAdvancedDateSelector, ICalendarTimezone } from '../../types';

export interface IAdvancedDateSelectDropdown extends ICalendarAdvancedDateSelector {
	/** (optional) The text field dropdown custom configurations */
	inputConfig?: Partial<ITextField>;
	/** (optional) The text field date format mask in ISO 8601 format */
	dateMask?: string;
	/** (optional) The dropdown position options */
	dropdownPositionOptions?: Partial<ComputePositionConfig>;
}

export interface IAdvancedDateSelectDropdownEvents {
	/** Emitted when the calendar selector opens state changes */
	openStateChange: EventEmitter<boolean>;
	/** Emitted when a time selector is applied */
	timeApplied: EventEmitter<ITimeChange>;
}

export interface ITimeChange {
	time: ICalendarAdvanceTimeChange;
	timezone: ICalendarTimezone;
}
