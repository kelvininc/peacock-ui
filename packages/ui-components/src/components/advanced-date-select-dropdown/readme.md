# _kv-advanced-date-select-dropdown_

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-advanced-date-select-dropdown></kv-advanced-date-select-dropdown>

<!-- With selected absolute time -->
<kv-advanced-date-select-dropdown
	[selectedTime]="{
		type: ECalendarAdvanceTimeType.Relative,
		payload: {
			key: 'last-24-h',
			range: ['2021-12-01T00:00:00', '2021-12-05T23:59:59Z']
		}
	}"
></kv-advanced-date-select-dropdown>

<!-- With selected relative time -->
<kv-advanced-date-select-dropdown
	[selectedTime]="{
		type: ECalendarAdvanceTimeType.Absolute,
		payload: {
			key: '2021-12-01T00:00:00Z#2021-12-05T23:59:59Z',
			range: ['2021-12-01T00:00:00', '2021-12-05T23:59:59Z']
		}
	}"
></kv-advanced-date-select-dropdown>

<!-- With selected timezone time -->
<kv-advanced-date-select-dropdown selectedTimezone="Europe/Lisbon"></kv-advanced-date-select-dropdown>
```


### React

```tsx
import React from 'react';

import { KvAdvancedDateSelectDropdown, ECalendarAdvanceTimeType } from '@kelvininc/react-ui-components';

export const KvAdvancedDateExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvAdvancedDateSelectDropdown />

		{/*-- With selected relative time --*/}
		<KvAdvancedDateSelectDropdown
			selectedTime={{
				type: ECalendarAdvanceTimeType.Relative,
				payload: {
					key: 'last-24-h',
					range: ['2021-12-01T00:00:00', '2021-12-05T23:59:59Z']
				}
			}}
		/>

		{/*-- With selected absolute time --*/}
		<KvAdvancedDateSelectDropdown
			selectedTime={{
				type: ECalendarAdvanceTimeType.Absolute,
				payload: {
					key: '2021-12-01T00:00:00Z#2021-12-05T23:59:59Z',
					range: ['2021-12-01T00:00:00', '2021-12-05T23:59:59Z']
				}
			}}
		/>

		{/*-- With selected timezone --*/}
		<KvAdvancedDateSelectDropdown selectedTimezone="Europe/Lisbon" />
	</>
);
```



## Properties

| Property                  | Attribute           | Description                                                                                                                                                                             | Type                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | Default                             |
| ------------------------- | ------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------- |
| `absoluteTimeConfig`      | --                  | (optional) The absolute time custom configurations                                                                                                                                      | `ICalendarAdvanceAbsoluteTimeConfig`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | `undefined`                         |
| `customClass`             | `custom-class`      | (optional) Additional classes to apply for custom CSS. If multiple classes are provided they should be separated by spaces. It is also valid to provide CssClassMap with boolean logic. | `CssClassMap \| string \| string[]`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `''`                                |
| `dateMask`                | `date-mask`         | (optional) The text field date format mask in ISO 8601 format                                                                                                                           | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | `undefined`                         |
| `disabled`                | `disabled`          | (optional) If `true` clicking on the input will not open the calendar dropdown. Default: false                                                                                          | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | `false`                             |
| `dropdownPositionOptions` | --                  | (optional) The dropdown position options                                                                                                                                                | `{ placement?: Placement; strategy?: Strategy; middleware?: (false \| { name: string; options?: any; fn: (state: { x: number; y: number; initialPlacement: Placement; platform: Platform; placement: Placement; strategy: Strategy; middlewareData: MiddlewareData; rects: ElementRects; elements: Elements; }) => Promisable<MiddlewareReturn>; })[]; platform?: Platform; }`                                                                                                                                                                                                               | `DEFAULT_DROPDOWN_POSITION_OPTIONS` |
| `inputConfig`             | --                  | (optional) The text field dropdown custom configurations                                                                                                                                | `{ type?: EInputFieldType; label?: string; icon?: EIconName \| EOtherIconName; inputName?: string; examples?: string[]; placeholder?: string; maxLength?: number; minLength?: number; max?: string \| number; min?: string \| number; step?: string \| number; size?: EComponentSize; disabled?: boolean; required?: boolean; loading?: boolean; state?: EValidationState; helpText?: string \| string[]; value?: string \| number; readonly?: boolean; forcedFocus?: boolean; highlighted?: boolean; tooltipConfig?: Partial<ITooltip>; useInputMask?: boolean; inputMaskRegex?: string; }` | `{}`                                |
| `relativeTimeConfig`      | --                  | (optional) The absolute time custom configurations                                                                                                                                      | `ICalendarAdvanceRelativeTimeConfig`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | `undefined`                         |
| `selectedTime`            | --                  | (optional) The selected time range                                                                                                                                                      | `ICalendarAdvanceSelectedTime`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | `undefined`                         |
| `selectedTimezone`        | `selected-timezone` | (optional) Currently selected timezone name                                                                                                                                             | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | `getDefaultTimezone()`              |
| `timezones`               | --                  | (optional) List with all selectable timezones                                                                                                                                           | `string[]`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | `getTimezonesNames()`               |


## Events

| Event                 | Description                                           | Type                       |
| --------------------- | ----------------------------------------------------- | -------------------------- |
| `dropdownStateChange` | Emitted when the calendar dropdown open state changes | `CustomEvent<boolean>`     |
| `timeApplied`         | Emitted when a time selector is applied               | `CustomEvent<ITimeChange>` |


## Dependencies

### Depends on

- [kv-dropdown](../dropdown)
- [kv-calendar-advanced-date-selector](../calendar-advanced-date-selector)
- [kv-action-button-text](../action-button-text)

### Graph
```mermaid
graph TD;
  kv-advanced-date-select-dropdown --> kv-dropdown
  kv-advanced-date-select-dropdown --> kv-calendar-advanced-date-selector
  kv-advanced-date-select-dropdown --> kv-action-button-text
  kv-dropdown --> kv-dropdown-base
  kv-dropdown --> kv-text-field
  kv-dropdown --> kv-icon
  kv-dropdown-base --> kv-portal
  kv-text-field --> kv-tooltip
  kv-text-field --> kv-form-label
  kv-text-field --> kv-icon
  kv-text-field --> kv-form-help-text
  kv-tooltip --> kv-portal
  kv-tooltip --> kv-tooltip-text
  kv-form-help-text --> kv-icon
  kv-calendar-advanced-date-selector --> kv-range-dates-select-dropdown
  kv-calendar-advanced-date-selector --> kv-select
  kv-calendar-advanced-date-selector --> kv-select-option
  kv-calendar-advanced-date-selector --> kv-single-select-dropdown
  kv-range-dates-select-dropdown --> kv-dropdown-base
  kv-range-dates-select-dropdown --> kv-text-field
  kv-range-dates-select-dropdown --> kv-icon
  kv-range-dates-select-dropdown --> kv-calendar-range-dates-selector
  kv-calendar-range-dates-selector --> kv-calendar
  kv-calendar --> kv-action-button-icon
  kv-calendar --> kv-calendar-day
  kv-action-button-icon --> kv-action-button
  kv-action-button-icon --> kv-icon
  kv-action-button-icon --> kv-badge
  kv-select --> kv-search
  kv-search --> kv-text-field
  kv-search --> kv-icon
  kv-select-option --> kv-checkbox
  kv-select-option --> kv-select-option
  kv-checkbox --> kv-icon
  kv-single-select-dropdown --> kv-select-option
  kv-single-select-dropdown --> kv-dropdown
  kv-single-select-dropdown --> kv-select
  kv-single-select-dropdown --> kv-select-shortcuts-label
  kv-select-shortcuts-label --> kv-icon
  kv-action-button-text --> kv-action-button
  kv-action-button-text --> kv-icon
  style kv-advanced-date-select-dropdown fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


