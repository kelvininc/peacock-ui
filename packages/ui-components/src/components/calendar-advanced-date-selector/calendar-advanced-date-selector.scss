@import '../../assets/styles/globals';

.advanced-date-selector {
	background-color: kv-color('neutral-7');
	border: 0.5px solid kv-color('neutral-6');
	border-radius: 4px;
	width: 544px;
}

.selector {
	border: 0.5px solid kv-color('neutral-6');
	padding: $spacing-4x;
}

.date-selector {
	display: flex;
	flex: 1;

	.absolute-date-selector,
	.relative-date-selector {
		flex: 1;
		padding: $spacing-4x;
	}

	.date-selector-label {
		@include kv-font-span-semibold;
		color: kv-color('neutral-2');
		border-radius: 4px;
		margin-bottom: $spacing-5x;
	}

	.relative-date-selector kv-select::part(select) {
		--select-border: none;
		--select-border-radius: 0;
	}

	.absolute-date-selector .start-single-date-select-dropdown,
	.absolute-date-selector .end-single-date-select-dropdown {
		kv-text-field {
			--text-color-input-default: #{kv-color('neutral-4')};
			--text-color-input-focused: #{kv-color('neutral-0')};
			--text-color-placeholder-focused: #{kv-color('neutral-0')};
		}
	}
}

.timezone-selector kv-single-select-dropdown {
	--dropdown-max-height: 300px;

	&::part(input) {
		--text-color-input-default: #{kv-color('neutral-4')};
	}
}
