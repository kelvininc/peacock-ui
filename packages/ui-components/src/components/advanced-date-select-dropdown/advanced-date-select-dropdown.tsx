import { Component, Element, Event, EventEmitter, Fragment, Host, Prop, State, Watch, h } from '@stencil/core';
import { DEFAULT_DATE_INPUT_CONFIG, DEFAULT_DROPDOWN_POSITION_OPTIONS } from './advanced-date-select-dropdown.config';
import {
	EActionButtonType,
	ECalendarAdvanceTimeType,
	ICalendarAdvanceAbsoluteTimeConfig,
	ICalendarAdvanceRelativeTimeConfig,
	ICalendarAdvanceSelectedTime,
	ICalendarAdvanceTime,
	ITextField,
	SelectedRange
} from '../../types';
import { IAdvancedDateSelectDropdown, IAdvancedDateSelectDropdownEvents, ITimeChange } from './advanced-date-select-dropdown.types';
import {
	formatAbsoluteSelectedTime,
	getRelativeTimeLabel,
	getTimeRange,
	isAbsoluteTimeSelected,
	isRelativeTimeSelected,
	isTimeSelected
} from './advanced-date-select-dropdown.helper';
import { formatTimezoneName, fromDatesRangeKey, getDefaultTimezone, getTimezoneOffset, getTimezonesNames } from '../../utils/date.helper';
import { isEqual, merge } from 'lodash-es';

import { ComputePositionConfig } from '@floating-ui/dom';

@Component({
	tag: 'kv-advanced-date-select-dropdown',
	styleUrl: 'advanced-date-select-dropdown.scss',
	shadow: true
})
export class KvAdvancedDateSelectDropdown implements IAdvancedDateSelectDropdown, IAdvancedDateSelectDropdownEvents {
	/** @inheritdoc */
	@Prop({ reflect: false }) inputConfig?: Partial<ITextField> = {};
	/** @inheritdoc */
	@Prop({ reflect: false }) selectedTime?: ICalendarAdvanceSelectedTime;
	/** @inheritdoc */
	@Prop({ reflect: false }) absoluteTimeConfig?: ICalendarAdvanceAbsoluteTimeConfig;
	/** @inheritdoc */
	@Prop({ reflect: false }) relativeTimeConfig?: ICalendarAdvanceRelativeTimeConfig;
	/** @inheritdoc */
	@Prop({ reflect: false }) selectedTimezone?: string = getDefaultTimezone();
	/** @inheritdoc */
	@Prop({ reflect: false }) timezones?: string[] = getTimezonesNames();
	/** @inheritdoc */
	@Prop({ reflect: false }) dateMask?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) dropdownPositionOptions?: Partial<ComputePositionConfig> = DEFAULT_DROPDOWN_POSITION_OPTIONS;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled?: boolean = false;

	/** @inheritdoc */
	@Event({ bubbles: false }) openStateChange: EventEmitter<boolean>;
	/** @inheritdoc */
	@Event() timeApplied: EventEmitter<ITimeChange>;

	@State() internalSelectedTime: ICalendarAdvanceSelectedTime | undefined;
	@State() internalSelectedTimezone: string;
	@State() isDropdownOpen: boolean = false;

	@Element() element: HTMLKvAdvancedDateSelectDropdownElement;

	@Watch('selectedTime')
	handleSelectedTimeChange(newSelectedTime: ICalendarAdvanceSelectedTime | undefined) {
		this.internalSelectedTime = newSelectedTime;
	}

	handleSelectedTimezoneChange(newTimezone: string) {
		this.internalSelectedTimezone = newTimezone;
	}

	componentWillLoad() {
		this.handleSelectedTimeChange(this.selectedTime);
		this.handleSelectedTimezoneChange(this.selectedTimezone);
	}

	private getTextFieldTooltip = (): string | undefined => {
		if (this.selectedTime !== undefined) {
			if (this.timezones.length > 0) {
				if (this.selectedTimezone !== undefined) {
					return formatTimezoneName(this.selectedTimezone);
				}

				const defaultTimezone = getDefaultTimezone();
				if (this.timezones.includes(defaultTimezone)) {
					return formatTimezoneName(defaultTimezone);
				}
			} else {
				return this.getFormattedSelectedTime();
			}
		}
	};

	private getFormattedSelectedTime = (): string | undefined => {
		if (this.selectedTime !== undefined) {
			if (this.selectedTime.type === ECalendarAdvanceTimeType.Relative) {
				return getRelativeTimeLabel(this.selectedTime.key as string | undefined);
			}

			if (this.selectedTime.type === ECalendarAdvanceTimeType.Absolute) {
				const [startDate, endDate] = fromDatesRangeKey(this.selectedTime.key) as SelectedRange;

				if (startDate !== undefined && endDate !== undefined) {
					return formatAbsoluteSelectedTime(startDate, endDate, this.dateMask);
				}
			}
		}
	};

	private onRelativeTimeChange = ({ detail: newRelativeTime }: CustomEvent<ICalendarAdvanceTime>) => {
		this.internalSelectedTime = {
			type: ECalendarAdvanceTimeType.Relative,
			key: newRelativeTime.key
		};
	};

	private onAbsoluteTimeChange = ({ detail: newAbsoluteTime }: CustomEvent<ICalendarAdvanceTime>) => {
		this.internalSelectedTime = {
			type: ECalendarAdvanceTimeType.Absolute,
			key: newAbsoluteTime.key
		};
	};

	private onTimezoneChange = ({ detail: newTimezone }: CustomEvent<string>) => {
		this.internalSelectedTimezone = newTimezone;
	};

	private onClickApply = () => {
		this.timeApplied.emit({
			time: {
				type: this.internalSelectedTime.type,
				payload: {
					key: this.internalSelectedTime.key,
					range: getTimeRange(this.internalSelectedTime, this.relativeTimeConfig?.options, this.internalSelectedTimezone)
				}
			},
			timezone: {
				name: this.internalSelectedTimezone,
				offset: getTimezoneOffset(this.internalSelectedTimezone)
			}
		});
		this.closeDropdown();
	};

	private onClickCancel = () => {
		this.internalSelectedTime = this.selectedTime;
		this.internalSelectedTimezone = this.selectedTimezone;

		this.closeDropdown();
	};

	private onDropdownChange = ({ detail: openState }: CustomEvent<boolean>) => {
		this.isDropdownOpen = openState;
	};

	private closeDropdown = () => {
		this.isDropdownOpen = !this.isDropdownOpen;
	};

	public isApplyDisabled = (): boolean => {
		// if there's no internal selected time, apply should be disabled
		if (!isTimeSelected(this.internalSelectedTime)) {
			return true;
		}

		// if the internal selected time is not the same as the external selected time, apply should be enabled
		if (!isEqual(this.internalSelectedTime, this.selectedTime)) {
			return false;
		}

		// if the selected time is the same but the internal timezone is different from the external timezone, apply should be enabled
		if (this.internalSelectedTimezone !== this.selectedTimezone) {
			return false;
		}

		return true;
	};

	public getInputConfig = (): Partial<ITextField> => {
		return merge({}, DEFAULT_DATE_INPUT_CONFIG, this.inputConfig, { value: this.getFormattedSelectedTime(), tooltipConfig: { text: this.getTextFieldTooltip() } });
	};

	render() {
		return (
			<Host>
				<div
					class={{
						'advanced-date-select-dropdown': true,
						'advanced-date-select-dropdown--selected-relative': isRelativeTimeSelected(this.selectedTime),
						'advanced-date-select-dropdown--selected-absolute': isAbsoluteTimeSelected(this.selectedTime)
					}}
				>
					<kv-dropdown
						isOpen={this.isDropdownOpen}
						onOpenStateChange={this.onDropdownChange}
						inputConfig={this.getInputConfig()}
						options={this.dropdownPositionOptions}
						disabled={this.disabled}
					>
						<div id="calendar" class="calendar-container">
							<kv-calendar-advanced-date-selector
								selectedTime={this.internalSelectedTime}
								absoluteTimeConfig={this.absoluteTimeConfig}
								relativeTimeConfig={this.relativeTimeConfig}
								selectedTimezone={this.internalSelectedTimezone}
								timezones={this.timezones}
								onRelativeTimeChange={this.onRelativeTimeChange}
								onAbsoluteTimeChange={this.onAbsoluteTimeChange}
								onTimezoneChange={this.onTimezoneChange}
							/>
							<div class="footer">
								<div class="selected-time">
									{isTimeSelected(this.selectedTime) && (
										<Fragment>
											<div class="label">Range in use:</div>
											<div class="value">{this.getFormattedSelectedTime()}</div>
										</Fragment>
									)}
								</div>
								<div class="actions">
									<kv-action-button-text type={EActionButtonType.Tertiary} text="Cancel" onClickButton={this.onClickCancel} />
									<kv-action-button-text type={EActionButtonType.Primary} text="Apply" disabled={this.isApplyDisabled()} onClickButton={this.onClickApply} />
								</div>
							</div>
						</div>
					</kv-dropdown>
				</div>
			</Host>
		);
	}
}
