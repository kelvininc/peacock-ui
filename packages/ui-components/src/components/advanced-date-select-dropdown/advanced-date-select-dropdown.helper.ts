import { isEmpty } from 'lodash-es';
import { RelativeTimeOption, SelectedRange } from '../../types';
import { formatDateTime, formatForTimezone, fromDatesRangeKey } from '../../utils/date.helper';
import { DEFAULT_RELATIVE_TIME_OPTIONS } from '../calendar-advanced-date-selector/calendar-advanced-date-selector.config';
import { getDatesRangeFromRelativeOption } from '../calendar-advanced-date-selector/calendar-advanced-date-selector.helper';
import { ICalendarAdvanceSelectedTime, ECalendarAdvanceTimeType } from '../calendar-advanced-date-selector/calendar-advanced-date-selector.types';

export const formatAbsoluteSelectedTime = (startDate: string, endDate: string, mask?: string): string => {
	return `${formatDateTime(startDate, mask)} to ${formatDateTime(endDate, mask)}`;
};

export const getRelativeTimeLabel = (relativeTimeValue: string | undefined, relativeTimeOptions: RelativeTimeOption[] = DEFAULT_RELATIVE_TIME_OPTIONS): string | undefined =>
	relativeTimeOptions.find(option => option.value === relativeTimeValue)?.label ?? relativeTimeValue;

export const isTimeSelected = (time: ICalendarAdvanceSelectedTime | undefined): boolean => {
	return isAbsoluteTimeSelected(time) || isRelativeTimeSelected(time);
};

export const isAbsoluteTimeSelected = (time: ICalendarAdvanceSelectedTime | undefined): boolean => {
	if (time && time.type === ECalendarAdvanceTimeType.Absolute) {
		const [startDate, endDate] = fromDatesRangeKey(time.key) as SelectedRange;

		return !isEmpty(startDate) && !isEmpty(endDate);
	}

	return false;
};

export const getTimeRange = (
	time: ICalendarAdvanceSelectedTime | undefined,
	relativeTimeOptions: RelativeTimeOption[] = DEFAULT_RELATIVE_TIME_OPTIONS,
	timezone: string
): SelectedRange => {
	if (time === undefined) {
		return [];
	}

	if (time.type === ECalendarAdvanceTimeType.Relative) {
		return getDatesRangeFromRelativeOption(time.key as string | undefined, relativeTimeOptions, timezone);
	}

	if (time.type === ECalendarAdvanceTimeType.Absolute) {
		const [startDate, endDate] = fromDatesRangeKey(time.key);

		return [startDate, endDate].map(date => formatForTimezone(timezone, date)) as SelectedRange;
	}
};

export const isRelativeTimeSelected = (time: ICalendarAdvanceSelectedTime | undefined): boolean => time && time.key && time.type === ECalendarAdvanceTimeType.Relative;
