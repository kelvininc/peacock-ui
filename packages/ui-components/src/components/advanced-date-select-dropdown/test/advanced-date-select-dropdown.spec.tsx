jest.mock('../../../utils/date.helper', () => ({
	...jest.requireActual('../../../utils/date.helper'),
	getDefaultTimezone: () => 'Europe/Lisbon'
}));

import { h, SpecPage } from '@stencil/core/internal';
import { newSpecPage } from '@stencil/core/testing';
import { ECalendarAdvanceTimeType } from '../../calendar-advanced-date-selector/calendar-advanced-date-selector.types';
import { ITextField } from '../../text-field/text-field.types';
import { KvAdvancedDateSelectDropdown } from '../advanced-date-select-dropdown';

describe('Advance Date Select Dropdown (unit tests)', () => {
	let page: SpecPage;
	let component: KvAdvancedDateSelectDropdown;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvAdvancedDateSelectDropdown],
				html: '<kv-advanced-date-select-dropdown></kv-advanced-date-select-dropdown>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when `isApplyDisabled` is called', () => {
		describe("and there's no internal time selected", () => {
			let actualResult: boolean;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					html: '<kv-advanced-date-select-dropdown></kv-advanced-date-select-dropdown>'
				});
				component = page.rootInstance;

				actualResult = component.isApplyDisabled();
			});

			it('is should return true', () => {
				expect(actualResult).toBe(true);
			});
		});

		describe('and internal time is not equals to prop time', () => {
			let actualResult: boolean;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					template: () => (
						<kv-advanced-date-select-dropdown
							selectedTime={{
								type: ECalendarAdvanceTimeType.Relative,
								key: 'last-24-h'
							}}
						/>
					)
				});
				component = page.rootInstance;
				component.internalSelectedTime = {
					type: ECalendarAdvanceTimeType.Absolute,
					key: '2020-04-01T00:00:00Z#2020-04-05T23:59:59Z'
				};

				actualResult = component.isApplyDisabled();
			});

			it('is should return false', () => {
				expect(actualResult).toBe(false);
			});
		});

		describe('and internal timezone is not equals to prop timezone', () => {
			let actualResult: boolean;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					template: () => (
						<kv-advanced-date-select-dropdown
							selectedTime={{
								type: ECalendarAdvanceTimeType.Relative,
								key: 'last-24-h'
							}}
							selectedTimezone="Europe/Lisbon"
						/>
					)
				});
				component = page.rootInstance;
				component.internalSelectedTimezone = 'Africa/Abidjan';

				actualResult = component.isApplyDisabled();
			});

			it('is should return false', () => {
				expect(actualResult).toBe(false);
			});
		});

		describe('time is selected an internal timezone and time are equals to the props', () => {
			let actualResult: boolean;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					template: () => (
						<kv-advanced-date-select-dropdown
							selectedTimezone="Europe/Lisbon"
							selectedTime={{
								type: ECalendarAdvanceTimeType.Relative,
								key: 'last-24-h'
							}}
						/>
					)
				});
				component = page.rootInstance;
				component.internalSelectedTimezone = 'Europe/Lisbon';
				component.internalSelectedTime = {
					type: ECalendarAdvanceTimeType.Relative,
					key: 'last-24-h'
				};

				actualResult = component.isApplyDisabled();
			});

			it('is should return true', () => {
				expect(actualResult).toBe(true);
			});
		});
	});

	describe('when `getInputConfig` is called', () => {
		describe('and the time is no selected', () => {
			let actualResult: Partial<ITextField>;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					template: () => <kv-advanced-date-select-dropdown />
				});
				component = page.rootInstance;

				actualResult = component.getInputConfig();
			});

			it('should return a config with undefined as the value', () => {
				expect(actualResult.value).toBeUndefined();
			});

			it('should return a config with undefined as the tooltip', () => {
				expect(actualResult.tooltipConfig.text).toBeUndefined();
			});
		});

		describe('and the selected time is relative', () => {
			let actualResult: Partial<ITextField>;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					template: () => (
						<kv-advanced-date-select-dropdown
							selectedTime={{
								type: ECalendarAdvanceTimeType.Relative,
								key: 'last-24-h'
							}}
						/>
					)
				});
				component = page.rootInstance;

				actualResult = component.getInputConfig();
			});

			it('should return a config with the relative time label as the value', () => {
				expect(actualResult.value).toBe('Last 24 hours');
			});
		});

		describe('and the selected time is absolute', () => {
			let actualResult: Partial<ITextField>;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					template: () => (
						<kv-advanced-date-select-dropdown
							selectedTime={{
								type: ECalendarAdvanceTimeType.Absolute,
								key: '2020-04-01T00:00:00Z#2020-04-05T23:59:59Z'
							}}
						/>
					)
				});
				component = page.rootInstance;

				actualResult = component.getInputConfig();
			});

			it('should return a config with the absolute time range as the value', () => {
				expect(actualResult.value).toBe('2020-04-01 00:00:00 to 2020-04-05 23:59:59');
			});
		});

		describe('and the timezone is not selected', () => {
			let actualResult: Partial<ITextField>;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					template: () => (
						<kv-advanced-date-select-dropdown
							selectedTime={{
								type: ECalendarAdvanceTimeType.Absolute,
								key: '2020-04-01T00:00:00Z#2020-04-05T23:59:59Z'
							}}
						/>
					)
				});
				component = page.rootInstance;

				actualResult = component.getInputConfig();
			});

			it('should return a config with default timezone as the tooltip', () => {
				expect(actualResult.tooltipConfig.text).toBe('(+01:00) Europe/Lisbon');
			});
		});

		describe('and the timezone is selected', () => {
			let actualResult: Partial<ITextField>;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvAdvancedDateSelectDropdown],
					template: () => (
						<kv-advanced-date-select-dropdown
							selectedTime={{
								type: ECalendarAdvanceTimeType.Absolute,
								key: '2020-04-01T00:00:00Z#2020-04-05T23:59:59Z'
							}}
							selectedTimezone="Africa/Abidjan"
						/>
					)
				});
				component = page.rootInstance;

				actualResult = component.getInputConfig();
			});

			it('should return a config with selected timezone as the tooltip', () => {
				expect(actualResult.tooltipConfig.text).toBe('(+00:00) Africa/Abidjan');
			});
		});
	});
});
