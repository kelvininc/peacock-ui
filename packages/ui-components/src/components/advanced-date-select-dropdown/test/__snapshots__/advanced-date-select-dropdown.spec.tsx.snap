// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Advance Date Select Dropdown (unit tests) when uses default props should match the snapshot 1`] = `
<kv-advanced-date-select-dropdown class="advanced-date-select-dropdown" custom-class="">
  <kv-dropdown clickoutsideclose="" zindex="9004">
    <div class="advanced-date-select-calendar-container" id="calendar">
      <kv-calendar-advanced-date-selector selectedtimezone="Europe/Lisbon"></kv-calendar-advanced-date-selector>
      <div class="footer">
        <div class="selected-time"></div>
        <div class="actions">
          <kv-action-button-text text="Cancel" type="tertiary"></kv-action-button-text>
          <kv-action-button-text disabled="" text="Apply" type="primary"></kv-action-button-text>
        </div>
      </div>
    </div>
  </kv-dropdown>
</kv-advanced-date-select-dropdown>
`;
